drivers/platform/stm32/gpio.c

```diff
@@ -0,0 +1,320 @@
+#include "util.h"
+#include "gpio.h"
+#include "gpio_extra.h"
+
+/**
+ * @brief stm32 platform specific GPIO platform ops structure
+ */
+const struct gpio_platform_ops stm32_gpio_platform_ops = {
+	.gpio_ops_get = &stm32_gpio_get,
+	.gpio_ops_get_optional = &stm32_gpio_get_optional,
+	.gpio_ops_remove = &stm32_gpio_remove,
+	.gpio_ops_direction_input = &stm32_gpio_direction_input,
+	.gpio_ops_direction_output = &stm32_gpio_direction_output,
+	.gpio_ops_get_direction = &stm32_gpio_get_direction,
+	.gpio_ops_set_value = &stm32_gpio_set_value,
+	.gpio_ops_get_value = &stm32_gpio_get_value,
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Prepare the GPIO decriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - The structure that contains the GPIO parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t _gpio_init(struct gpio_desc *desc,
+		   const struct gpio_init_param *param)
+{
+	int32_t ret = 0;
+	struct stm32_gpio_desc *extra = desc->extra;
+	struct stm32_gpio_desc *pextra = param->extra;
+	GPIO_InitTypeDef gis;
+
+	if (!param)
+		return -EINVAL;
+
+	/* enable gpio port in RCC */
+	if (pextra->port == GPIOA)
+		__HAL_RCC_GPIOA_CLK_ENABLE();
+#ifdef GPIOB
+	else if (pextra->port == GPIOB)
+		__HAL_RCC_GPIOB_CLK_ENABLE();
+#endif
+#ifdef GPIOC
+	else if (pextra->port == GPIOC)
+		__HAL_RCC_GPIOC_CLK_ENABLE();
+#endif
+#ifdef GPIOD
+	else if (pextra->port == GPIOD)
+		__HAL_RCC_GPIOD_CLK_ENABLE();
+#endif
+#ifdef GPIOE
+	else if (pextra->port == GPIOE)
+		__HAL_RCC_GPIOE_CLK_ENABLE();
+#endif
+#ifdef GPIOF
+	else if (pextra->port == GPIOF)
+		__HAL_RCC_GPIOF_CLK_ENABLE();
+#endif
+#ifdef GPIOG
+	else if (pextra->port == GPIOG)
+		__HAL_RCC_GPIOG_CLK_ENABLE();
+#endif
+#ifdef GPIOH
+	else if (pextra->port == GPIOH)
+		__HAL_RCC_GPIOH_CLK_ENABLE();
+#endif
+	else
+		return -EINVAL;
+
+	/* copy the settings to gpio descriptor */
+	desc->number = param->number;
+	extra->port = pextra->port;
+	extra->mode = pextra->mode;
+	extra->pull = pextra->pull;
+	extra->speed = pextra->speed;
+
+	/* configure gpio with user configuration */
+	gis.Pin = BIT(param->number);
+	gis.Mode = extra->mode;
+	gis.Pull = extra->pull;
+	gis.Speed = extra->speed;
+	HAL_GPIO_Init(extra->port, &gis);
+
+	return ret;
+}
+
+/**
+ * @brief Obtain the GPIO decriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - GPIO initialization parameters
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_gpio_get(struct gpio_desc **desc,
+		       const struct gpio_init_param *param)
+{
+	struct gpio_desc *descriptor;
+	struct stm32_gpio_desc *extra;
+	int32_t ret;
+
+	descriptor = (struct gpio_desc *)malloc(sizeof(*descriptor));
+	if (!descriptor) {
+		ret = -ENOMEM;
+		goto error;
```

Comment:
no need to go to error. Will free a NULL pointer.
Same for extra

Reply:
freeing null pointer is ok